File: csharp/src/Vista.SDK/Locations.cs
using System.Diagnostics;
using System.Diagnostics.CodeAnalysis;
using System.Globalization;
using Vista.SDK.Internal;

namespace Vista.SDK;

public enum LocationGroup
{
    Number,
    Side,
    Vertical,
    Transverse,
    Longitudinal
}

public readonly record struct Location
{
    public readonly string Value { get; }

    internal Location(string value)
    {
        Value = value;
    }

    public override readonly string ToString() => Value;

    public static implicit operator string(Location n) => n.Value;
}

public sealed class Locations
{
    private readonly char[] _locationCodes;
    private readonly List<RelativeLocation> _relativeLocations;
    internal Dictionary<char, LocationGroup> _reversedGroups;

    public VisVersion VisVersion { get; }

    // This is if we need Code, Name, Definition in Frontend UI
    public IReadOnlyList<RelativeLocation> RelativeLocations => _relativeLocations;

    public IReadOnlyDictionary<LocationGroup, IReadOnlyList<RelativeLocation>> Groups;

    internal Locations(VisVersion version, LocationsDto dto)
    {
        VisVersion = version;

        _locationCodes = dto.Items.Select(d => d.Code).ToArray();
        _relativeLocations = new List<RelativeLocation>(dto.Items.Length);

        var groups = new Dictionary<LocationGroup, List<RelativeLocation>>(5);

        _reversedGroups = new Dictionary<char, LocationGroup>();

        foreach (var relativeLocationsDto in dto.Items)
        {
            var relativeLocation = new RelativeLocation(
                relativeLocationsDto.Code,
                relativeLocationsDto.Name,
                new Location(relativeLocationsDto.Code.ToString()),
                relativeLocationsDto.Definition
            );
            _relativeLocations.Add(relativeLocation);

            // Not interested in horizontal and vertical codes
            if (relativeLocationsDto.Code is 'H' or 'V')
                continue;

            var key = relativeLocationsDto.Code switch
            {
                'N' => LocationGroup.Number,
                'P' or 'C' or 'S' => LocationGroup.Side,
                'U' or 'M' or 'L' => LocationGroup.Vertical,
                'I' or 'O' => LocationGroup.Transverse,
                'F' or 'A' => LocationGroup.Longitudinal,
                _ => throw new Exception($"Unsupported code: {relativeLocationsDto.Code}")
            };
            if (!groups.ContainsKey(key))
                groups.Add(key, new());
            if (key == LocationGroup.Number)
                continue;
            _reversedGroups.Add(relativeLocationsDto.Code, key);
            groups.GetValueOrDefault(key)?.Add(relativeLocation);
        }

        Groups = groups.ToDictionary(g => g.Key, g => g.Value.ToArray() as IReadOnlyList<RelativeLocation>);
    }

    public Location Parse(string locationStr)
    {
        var errorBuilder = LocationParsingErrorBuilder.Empty;

        var span = locationStr is null ? ReadOnlySpan<char>.Empty : locationStr.AsSpan();
        if (!TryParseInternal(span, locationStr, out var location, ref errorBuilder))
            throw new ArgumentException($"Invalid value for location: {locationStr}, errors: {errorBuilder.Build()}");

        return location;
    }

    public Location Parse(ReadOnlySpan<char> locationStr)
    {
        var errorBuilder = LocationParsingErrorBuilder.Empty;

        if (!TryParseInternal(locationStr, null, out var location, ref errorBuilder))
            throw new ArgumentException(
                $"Invalid value for location: {locationStr.ToString()}, errors: {errorBuilder.Build()}"
            );

        return location;
    }

    public bool TryParse(string? value, out Location location)
    {
        var span = value is null ? ReadOnlySpan<char>.Empty : value.AsSpan();
        var errorBuilder = LocationParsingErrorBuilder.Empty;
        return TryParseInternal(span, value, out location, ref errorBuilder);
    }

    public bool TryParse(string? value, out Location location, out ParsingErrors errors)
    {
        var span = value is null ? ReadOnlySpan<char>.Empty : value.AsSpan();
        var errorBuilder = LocationParsingErrorBuilder.Empty;
        var result = TryParseInternal(span, value, out location, ref errorBuilder);
        errors = errorBuilder.Build();
        return result;
    }

    public bool TryParse(ReadOnlySpan<char> value, out Location location)
    {
        var errorBuilder = LocationParsingErrorBuilder.Empty;
        return TryParseInternal(value, null, out location, ref errorBuilder);
    }

    public bool TryParse(ReadOnlySpan<char> value, out Location location, out ParsingErrors errors)
    {
        var errorBuilder = LocationParsingErrorBuilder.Empty;
        var result = TryParseInternal(value, null, out location, ref errorBuilder);
        errors = errorBuilder.Build();
        return result;
    }

    private bool TryParseInternal(
        ReadOnlySpan<char> span,
        string? originalStr,
        out Location location,
        ref LocationParsingErrorBuilder errorBuilder
    )
    {
        location = default;

        if (span.IsEmpty)
        {
            AddError(
                ref errorBuilder,
                LocationValidationResult.NullOrWhiteSpace,
                "Invalid location: contains only whitespace"
            );
            return false;
        }

        if (span.IsWhiteSpace())
        {
            AddError(
                ref errorBuilder,
                LocationValidationResult.NullOrWhiteSpace,
                "Invalid location: contains only whitespace"
            );
            return false;
        }

        var originalSpan = span;

        int? prevDigitIndex = null;
        int? digitStartIndex = null;
        int? charsStartIndex = null;
        int? number = null;

        Debug.Assert(Enum.GetValues(typeof(LocationGroup)).Length == 5);
        var charDict = new LocationCharDict(stackalloc char?[4]);

        for (int i = 0; i < span.Length; i++)
        {
            ref readonly var ch = ref span[i];

            if (char.IsDigit(ch))
            {
                // First digit should be at index 0
                if (digitStartIndex is null && i != 0)
                {
                    AddError(
                        ref errorBuilder,
                        LocationValidationResult.Invalid,
                        $"Invalid location: numeric location should start before location code(s) in location: '{originalStr ?? originalSpan.ToString()}'"
                    );
                    return false;
                }
                // Other digits should neighbor the first
                if (prevDigitIndex is not null && prevDigitIndex != (i - 1))
                {
                    AddError(
                        ref errorBuilder,
                        LocationValidationResult.Invalid,
                        $"Invalid location: cannot have multiple separated digits in location: '{originalStr ?? originalSpan.ToString()}'"
                    );
                    return false;
                }

                if (digitStartIndex is null)
                {
                    number = ch - '0';
                    digitStartIndex = i;
                }
                else
                {
                    if (!TryParseInt(span, digitStartIndex.Value, i - digitStartIndex.Value + 1, out var num))
                    {
                        AddError(
                            ref errorBuilder,
                            LocationValidationResult.Invalid,
                            $"Invalid location: failed to parse numeric location: '{originalStr ?? originalSpan.ToString()}'"
                        );
                        return false;
                    }
                    number = num;
                }

                prevDigitIndex = i;
            }
            else
            {
                if (!_reversedGroups.TryGetValue(ch, out var group))
                {
                    var invalidChars = string.Join(
                        ",",
                        (originalStr ?? originalSpan.ToString())
                            .Where(c => !char.IsDigit(c) && (c == 'N' || !_locationCodes.Contains(c)))
                            .Select(c => $"'{c}'")
                    );
                    AddError(
                        ref errorBuilder,
                        LocationValidationResult.InvalidCode,
                        $"Invalid location code: '{originalStr ?? originalSpan.ToString()}' with invalid location code(s): {invalidChars}"
                    );
                    return false;
                }

                if (!charDict.TryAdd(group, ch, out var existingCh))
                {
                    AddError(
                        ref errorBuilder,
                        LocationValidationResult.Invalid,
                        $"Invalid location: Multiple '{Enum.GetName(typeof(LocationGroup), group)}' values. Got both '{existingCh}' and '{ch}' in '{originalStr ?? originalSpan.ToString()}'"
                    );
                    return false;
                }

                if (charsStartIndex is null)
                {
                    charsStartIndex = i;
                }
                else if (i > 0)
                {
                    ref readonly var prevCh = ref span[i - 1];
                    if (ch.CompareTo(prevCh) < 0)
                    {
                        AddError(
                            ref errorBuilder,
                            LocationValidationResult.InvalidOrder,
                            $"Invalid location: '{originalStr ?? originalSpan.ToString()}' not alphabetically sorted"
                        );
                        return false;
                    }
                }
            }
        }

        location = new Location(originalStr ?? originalSpan.ToString());
        return true;
    }

    static void AddError(ref LocationParsingErrorBuilder errorBuilder, LocationValidationResult name, string message)
    {
        if (!errorBuilder.HasError)
            errorBuilder = LocationParsingErrorBuilder.Create();
        errorBuilder.AddError(name, message);
    }

    internal static bool TryParseInt(ReadOnlySpan<char> span, int start, int length, out int number)
    {
#if NETCOREAPP3_1_OR_GREATER
        return int.TryParse(span.Slice(start, length), NumberStyles.None, CultureInfo.InvariantCulture, out number);
#else
        return int.TryParse(
            span.Slice(start, length).ToString(),
            NumberStyles.None,
            CultureInfo.InvariantCulture,
            out number
        );
#endif
    }
}

internal ref struct LocationCharDict
{
    private readonly Span<char?> _table;

    public LocationCharDict(Span<char?> table)
    {
        Debug.Assert(table.Length == Enum.GetValues(typeof(LocationGroup)).Length - 1);
        _table = table;
    }

    private ref char? this[LocationGroup key]
    {
        get
        {
            var index = (int)key - 1;
            if (index >= _table.Length)
                throw new Exception($"Unsupported code: {key}");

            return ref _table[index];
        }
    }

    public bool TryAdd(LocationGroup key, char value, [MaybeNullWhen(true)] out char? existingValue)
    {
        ref var v = ref this[key];
        if (v is not null)
        {
            existingValue = v;
            return false;
        }

        existingValue = null;
        v = value;
        return true;
    }
}

public sealed record RelativeLocation
{
    public char Code { get; }
    public string Name { get; }
    public string? Definition { get; }

    public Location Location { get; }

    internal RelativeLocation(char code, string name, Location location, string? definition)
    {
        Code = code;
        Name = name;
        Definition = definition;
        Location = location;
    }

    public override int GetHashCode() => Code.GetHashCode();

    public bool Equals(RelativeLocation? other) => Code == other?.Code;
}

public enum LocationValidationResult
{
    Invalid,
    InvalidCode,
    InvalidOrder,
    NullOrWhiteSpace,
    Valid
}

File: python/vista_sdk/Locations.py
from __future__ import annotations

from enum import Enum
from typing import Dict, List, Optional, Tuple, overload

from vista_sdk.LocationsDto import LocationsDto
from vista_sdk.VisVersions import VisVersion

from .internal.LocationParsingErrorBuilder import (
    LocationParsingErrorBuilder,
    LocationValidationResult,
    ParsingErrors,
)


class LocationGroup(Enum):
    NUMBER = 0
    SIDE = 1
    VERTICAL = 2
    TRANSVERSE = 3
    LONGITUDINAL = 4


class Location:
    def __init__(self, value: str):
        self._value = value

    @property
    def value(self) -> str:
        return self._value

    def __str__(self) -> str:
        return self._value

    def __repr__(self) -> str:
        return f"Location(value={self._value!r})"

    @staticmethod
    def from_string(s: str) -> Location:
        return Location(s)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, Location):
            return NotImplemented
        return self._value == other._value

    def __hash__(self) -> int:
        return hash(self._value)


class RelativeLocation:
    def __init__(
        self, code: str, name: str, location: Location, definition: Optional[str] = None
    ):
        self._code: str = code
        self._name: str = name
        self._location: Location = location
        self._definition: Optional[str] = definition

    @property
    def code(self) -> str:
        return self._code

    @property
    def name(self) -> str:
        return self._name

    @property
    def location(self) -> Location:
        return self._location

    @property
    def definition(self) -> Optional[str]:
        return self._definition

    def __hash__(self) -> int:
        return hash(self._code)

    def __eq__(self, other: object) -> bool:
        if not isinstance(other, RelativeLocation):
            return NotImplemented
        if isinstance(other, RelativeLocation):
            return self._code == other._code
        return NotImplemented


class Locations:
    def __init__(self, version: VisVersion, dto: LocationsDto):
        self.vis_version = version
        self._location_codes = [d.code for d in dto.items]
        self._relative_locations = []
        self._reversed_groups = {}
        groups: Dict[LocationGroup, List[RelativeLocation]] = {}

        for relative_locations_dto in dto.items:
            relative_location = RelativeLocation(
                relative_locations_dto.code,
                relative_locations_dto.name,
                Location(str(relative_locations_dto.code)),
                relative_locations_dto.definition,
            )
            self._relative_locations.append(relative_location)

            key = self.determine_group_by_code(relative_locations_dto.code)
            if key not in groups:
                groups[key] = []
            groups[key].append(relative_location)
            if key != LocationGroup.NUMBER:
                self._reversed_groups[relative_locations_dto.code] = key

        self._groups: Dict = {k: tuple(v) for k, v in groups.items()}

    def determine_group_by_code(self, code: str) -> LocationGroup:
        if code in "N":
            return LocationGroup.NUMBER
        elif code in "PCS":
            return LocationGroup.SIDE
        elif code in "UML":
            return LocationGroup.VERTICAL
        elif code in "IO":
            return LocationGroup.TRANSVERSE
        elif code in "FA":
            return LocationGroup.LONGITUDINAL
        else:
            raise Exception(f"Unsupported code: {code}")

    @property
    def relative_locations(self) -> List[RelativeLocation]:
        return self._relative_locations.copy()

    @property
    def groups(self) -> Dict[LocationGroup, List[RelativeLocation]]:
        return self._groups

    def parse(self, location_str: Optional[str]) -> Location:
        error_builder = LocationParsingErrorBuilder.create()
        location = None

        success, location = self.try_parse_internal(
            location_str if location_str else "", location_str, error_builder
        )
        if not success or not location:
            error_details = error_builder.build()
            raise ValueError(
                f"Invalid value for location: '{location_str}', errors: {error_details}"
            )

        return location

    @overload
    def try_parse(self, value: Optional[str]) -> Tuple[bool, Optional[Location]]: ...
    @overload
    def try_parse(
        self, value: Optional[Location]
    ) -> Tuple[bool, Optional[Location]]: ...

    def try_parse(
        self, value: Optional[str] | Optional[Location]
    ) -> Tuple[bool, Optional[Location]]:
        if isinstance(value, Location):
            return True, value
        if isinstance(value, str):
            error_builder = LocationParsingErrorBuilder.create()
            location = None

            if (
                self.try_parse_internal(value if value else "", value, error_builder)
                and value
            ):
                location = Location(value)
            else:
                return False, None
            return location is not None, location
        else:
            raise ValueError("Invalid value for location")

    def try_parse_with_errors(
        self, value: Optional[str]
    ) -> Tuple[bool, Optional[Location], ParsingErrors]:
        error_builder = LocationParsingErrorBuilder.create()
        location = None
        result = self.try_parse_internal(value if value else "", value, error_builder)
        errors = error_builder.build()
        if result and value:
            location = Location(value)
        return result[0], location, errors

    def try_parse_internal(
        self,
        value: str,
        original_str: Optional[str],
        error_builder: LocationParsingErrorBuilder,
    ) -> Tuple[bool, Optional[Location]]:
        location = None

        if not value:
            error_builder.add_error(
                LocationValidationResult.NULL_OR_WHITE_SPACE,
                "Invalid location: contains only whitespace",
            )
            return False, location

        if value.isspace():
            error_builder.add_error(
                LocationValidationResult.NULL_OR_WHITE_SPACE,
                "Invalid location: contains only whitespace",
            )
            return False, location

        original_span = value
        prev_digit_index = None
        digit_start_index = None
        char_dict: dict[LocationGroup, str] = {}

        assert len(LocationGroup) == 5  # noqa : S101;

        for i, ch in enumerate(value):
            if ch.isdigit():
                if digit_start_index is None and i != 0:
                    error_builder.add_error(
                        LocationValidationResult.INVALID,
                        "Invalid location: numeric location should start before"
                        "location code(s) in location:"
                        f"'{original_str or original_span}'",
                    )
                    return False, location
                if prev_digit_index is not None and prev_digit_index != (i - 1):
                    error_builder.add_error(
                        LocationValidationResult.INVALID,
                        "Invalid location: cannot have multiple separated"
                        f"digits in location: '{original_str or original_span}'",
                    )
                    return False, location
                if digit_start_index is None:
                    digit_start_index = i

                prev_digit_index = i
            else:
                group = self._reversed_groups.get(ch)
                if not group:
                    invalid_chars = ",".join(
                        set(
                            c
                            for c in original_str or original_span
                            if not c.isdigit() and c not in self._location_codes
                        )
                    )
                    error_builder.add_error(
                        LocationValidationResult.INVALID_CODE,
                        f"Invalid location code: '{original_str or original_span}'"
                        f"with invalid location code(s): {invalid_chars}",
                    )
                    return False, location

                if group in char_dict and char_dict[group] != ch:
                    error_builder.add_error(
                        LocationValidationResult.INVALID,
                        f"Invalid location: Multiple '{group}' values."
                        " Got both '{char_dict[group]}' and '{ch}' in "
                        f"'{original_str or original_span}'",
                    )
                    return False, location
                char_dict[group] = ch

        location = Location(original_str or original_span)
        return True, location

    @staticmethod
    def try_parse_int(span: str, start: int, length: int) -> Tuple[bool, int]:
        try:
            if start + length > len(span):
                return False, 0
            number = int(span[start : start + length])
            return True, number
        except ValueError:
            return False, 0


class LocationCharDict:
    def __init__(self, size: int):
        self._table = [None] * size

    def __getitem__(self, key: LocationGroup) -> Optional[str]:
        index = key.value - 1
        if index >= len(self._table):
            raise Exception(f"Unsupported code: {key}")
        return self._table[index]

    def __setitem__(self, key: LocationGroup, value) -> None:
        index = key.value - 1
        if index >= len(self._table):
            raise Exception(f"Unsupported code: {key}")
        self._table[index] = value

    def try_add(self, key: LocationGroup, value: str) -> Tuple[bool, Optional[str]]:
        current_value = self[key]
        if current_value is not None:
            return False, current_value
        self[key] = value
        return True, None
